File: src/main/kotlin/Task2.kt
package org.example

import kotlinx.coroutines.*

fun main() = runBlocking {
    val list1 = listOf(0.0, 4.0, 9.0, 25.0, 36.0, 49.0)
    val list2 = listOf(0.0, -1.0, 2.0, 3.0, 4.0, 5.0)

    val startTime = System.currentTimeMillis()

    val job1 = launch { println(asyncMap(list1, ::sqrt)) }
    val job2 = launch { try { println(asyncMap(list2, ::sqrt))} catch (e: Exception) {println(e.message)} }

    job1.join()
    job2.join()

    val endTime = System.currentTimeMillis()
    val elapsedTime = endTime - startTime
    println("\nTime: $elapsedTime ms")
}

suspend fun <T, R> asyncMap(
    items: List<T>,
    asyncTransform: suspend (T) -> Result<R>,
    resultHandler: (Result<R>) -> R = { it.getOrThrow() }
): List<R> {
    return coroutineScope {
        items.map { item ->
            async { asyncTransform(item) }
        }.awaitAll().map(resultHandler)
    }
}

suspend fun sqrt(num: Double): Result<Double> {
    delay(1000)
    if (num < 0) {
        return Result.failure(Exception("negative value: $num"))
    }
    return Result.success(kotlin.math.sqrt(num))

}

File: src/main/kotlin/Task2.js
const asyncMap = (array, asyncTransform) => {
    return Promise.all(array.map(item => asyncTransform(item)));
};


const testFunc = async (num) => {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            if (num < 0) {
                reject(new Error(`negative value: ${num}`));
            } else if (num % 2 === 0) {
                resolve(num * 2);
            } else {
                resolve(num ** 2);
            }
        }, 1000);
    });
};

const numbers1 = [1, 2, 3, 1, 4];
const numbers2 = [1, 2, 3, -1, 4];

const start = performance.now();

const promise1 = asyncMap(numbers1, testFunc)
    .then((result) => {
        console.log(result);
    })
    .catch((error) => {
        console.error('Error:', error.message);
    });

const promise2 = asyncMap(numbers2, testFunc)
    .then((result) => {
        console.log(result);
    })
    .catch((error) => {
        console.error('Error:', error.message);
    });

Promise.all([promise1, promise2]).then(() => {
    const end = performance.now();
    console.log(`Time: ${end - start} ms`);
});
